File: KotlinLibrary/library/src/main/java/com/lee/library/livedatabus/LiveDataBus.java
package com.lee.library.livedatabus;

import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Observer;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

/**
 * @author jv.lee
 * @date 2019/3/30
 * 事件总线
 */
public class LiveDataBus {

    /**
     * 消息通道
     */
    private Map<String, BusMutableLiveData<Object>> bus;
    private static LiveDataBus instance;

    private LiveDataBus() {
        bus = new HashMap<>();
    }

    public static LiveDataBus getInstance() {
        if (instance == null) {
            synchronized (LiveDataBus.class) {
                if (instance == null) {
                    instance = new LiveDataBus();
                }
            }
        }
        return instance;
    }

    public <T> MutableLiveData<T> getChannel(String target, Class<T> type) {
        if (!bus.containsKey(target)) {
            bus.put(target, new BusMutableLiveData<>());
        }
        return (MutableLiveData<T>) bus.get(target);
    }

    public MutableLiveData<Object> getChannel(String target) {
        return getChannel(target, Object.class);
    }

    private static class ObserverWrapper<T> implements Observer<T> {

        private Observer<T> observer;

        public ObserverWrapper(Observer<T> observer) {
            this.observer = observer;
        }

        @Override
        public void onChanged(@Nullable T t) {
            if (observer != null) {
//                if (isCallOnObserve()) {
//                    return;
//                }
                observer.onChanged(t);
            }
        }

        private boolean isCallOnObserve() {
            StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
            if (stackTrace != null && stackTrace.length > 0) {
                for (StackTraceElement element : stackTrace) {
                    if ("android.arch.lifecycle.LiveData".equals(element.getClassName()) &&
                            "observeForever".equals(element.getMethodName())) {
                        return true;
                    }
                }
            }
            return false;
        }
    }

    private static class BusMutableLiveData<T> extends MutableLiveData<T> {

        private Map<Observer, Observer> observerMap = new HashMap<>();

        @Override
        public void observe(@NonNull LifecycleOwner owner, @NonNull Observer<T> observer) {
            super.observe(owner, observer);
            try {
                hook(observer);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        @Override
        public void observeForever(@NonNull Observer<T> observer) {
            if (!observerMap.containsKey(observer)) {
                observerMap.put(observer, new ObserverWrapper(observer));
            }
            super.observeForever(observerMap.get(observer));
        }

        @Override
        public void removeObserver(@NonNull Observer<T> observer) {
            Observer realObserver = null;
            if (observerMap.containsKey(observer)) {
                realObserver = observerMap.remove(observer);
            } else {
                realObserver = observer;
            }
            super.removeObserver(realObserver);
        }

        private void hook(@NonNull Observer<T> observer) throws Exception {
            //get wrapper's version
            Class<LiveData> classLiveData = LiveData.class;
            Field fieldObservers = classLiveData.getDeclaredField("mObservers");
            fieldObservers.setAccessible(true);
            Object objectObservers = fieldObservers.get(this);
            Class<?> classObservers = objectObservers.getClass();
            Method methodGet = classObservers.getDeclaredMethod("get", Object.class);
            methodGet.setAccessible(true);
            Object objectWrapperEntry = methodGet.invoke(objectObservers, observer);
            Object objectWrapper = null;
            if (objectWrapperEntry instanceof Map.Entry) {
                objectWrapper = ((Map.Entry) objectWrapperEntry).getValue();
            }
            if (objectWrapper == null) {
                throw new NullPointerException("Wrapper can not be bull!");
            }
            Class<?> classObserverWrapper = objectWrapper.getClass().getSuperclass();
            Field fieldLastVersion = classObserverWrapper.getDeclaredField("mLastVersion");
            fieldLastVersion.setAccessible(true);
            //get livedata's version
            Field fieldVersion = classLiveData.getDeclaredField("mVersion");
            fieldVersion.setAccessible(true);
            Object objectVersion = fieldVersion.get(this);
            //set wrapper's version
            fieldLastVersion.set(objectWrapper, objectVersion);
        }
    }

    /**
     * 存储非激活事件的临时容器
     */
    private Map<String, Observer> tempMap = new HashMap<>();

    /**
     * 订阅通知
     * @param lifecycleOwner
     */
    public void injectBus(LifecycleOwner lifecycleOwner) {
        Class<? extends LifecycleOwner> aClass = lifecycleOwner.getClass();
        //获取当前类所有方法
        Method[] declaredMethods = aClass.getDeclaredMethods();
        for (Method method : declaredMethods) {
            //激活通知
            InjectBus injectBus = method.getAnnotation(InjectBus.class);
            if (injectBus != null) {
                String value = injectBus.value();
                boolean isActive = injectBus.isActive();

                Observer<Object> observer = o -> {
                    try {
                        method.invoke(lifecycleOwner, o);
                    } catch (IllegalAccessException e) {
                        e.printStackTrace();
                    } catch (InvocationTargetException e) {
                        e.printStackTrace();
                    }
                };

                //是否是激活状态通知
                if (isActive) {
                    LiveDataBus.getInstance().getChannel(value).observe(lifecycleOwner, observer);
                } else {
                    tempMap.put(value, observer);
                    LiveDataBus.getInstance().getChannel(value).observeForever(observer);
                }
            }

        }
    }

    /**
     * 取消订阅通知 (仅在使用非激活状态可通知模式 需要取消订阅)
     * @param lifecycleOwner
     */
    public void unInjectBus(LifecycleOwner lifecycleOwner) {
        Class<? extends LifecycleOwner> aClass = lifecycleOwner.getClass();
        //获取当前类所有方法
        Method[] declaredMethods = aClass.getDeclaredMethods();
        for (Method method : declaredMethods) {
            InjectBus injectBus = method.getAnnotation(InjectBus.class);
            if (injectBus != null) {
                String value = injectBus.value();
                boolean isActive = injectBus.isActive();

                //非活跃状态通知取消订阅
                if (!isActive) {
                    Observer remove = tempMap.remove(value);
                    if (remove != null) {
                        LiveDataBus.getInstance().getChannel(value).removeObserver(remove);
                    }
                }
            }
        }
    }

}

File: KotlinLibraryX/library/src/main/java/com/lee/library/livedatabus/LiveDataBus.kt
package com.lee.library.livedatabus

import android.app.Activity
import androidx.fragment.app.Fragment
import androidx.lifecycle.*
import com.lee.library.livedatabus.LiveDataBus.ObserverWrapper

/**
 * 事件总线
 * @author jv.lee
 * @date 2019/3/30
 */
@Suppress("UNCHECKED_CAST")
class LiveDataBus private constructor() {

    companion object {
        val instance by lazy { LiveDataBus() }
    }

    /**
     * 消息通道
     */
    private val bus: HashMap<String, BusMutableLiveData<Any>> = HashMap()

    /***
     * 注入的当前事件次数(根据注入次数控制解绑时机,处理多次监听相同事件直到全部需要解除时才移除bus中的事件)
     */
    private val injectSize: HashMap<String, Int> = HashMap()

    fun <T> getChannel(type: Class<T>): MutableLiveData<T> {
        if (!bus.containsKey(type.simpleName)) {
            bus[type.simpleName] = BusMutableLiveData()
        }
        return bus[type.simpleName] as MutableLiveData<T>
    }

    /**
     * 记录注入事件次数
     */
    private fun injectSizeAdd(type: Class<*>) {
        val count = injectSize[type.simpleName] ?: 0
        injectSize[type.simpleName] = 1 + count
    }

    /**
     * 移除事件次数 根据次数决定是否移除事件
     */
    private fun injectSizeRemove(type: Class<*>) {
        val tag = type.simpleName
        val count = injectSize[tag] ?: return
        if (count == 1) {
            injectSize.remove(tag)
            bus.remove(tag)
        } else {
            injectSize[tag] = count - 1
        }
    }

    private class ObserverWrapper<T>(private val observer: Observer<T>) : Observer<T> {

        override fun onChanged(t: T) {
            if (isCallOnObserve()) {
                return
            }
            observer.onChanged(t)
        }

        private fun isCallOnObserve(): Boolean {
            val stackTrace = Thread.currentThread().stackTrace
            stackTrace.takeIf { it.isNotEmpty() }?.forEach { element ->
                if ("android.arch.lifecycle.LiveData" == element.className && "observeForever" == element.methodName) {
                    return true
                }
            }
            return false
        }
    }

    private class BusMutableLiveData<T> : MutableLiveData<T>() {
        private val observerMap = hashMapOf<Observer<*>, Observer<*>>()

        override fun observe(owner: LifecycleOwner, observer: Observer<in T>) {
            super.observe(owner, observer)
            try {
                hook(observer)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        override fun observeForever(observer: Observer<in T>) {
            if (!observerMap.containsKey(observer)) {
                observerMap[observer] = ObserverWrapper(observer)
            }
            super.observeForever(observerMap[observer] as Observer<in T>)
        }

        override fun removeObserver(observer: Observer<in T>) {
            val realObserver: Observer<T> = if (observerMap.containsKey(observer)) {
                (observerMap.remove(observer) as Observer<T>)
            } else {
                observer as Observer<T>
            }
            super.removeObserver(realObserver)
        }

        private fun hook(observer: Observer<in T>) {
            // get wrapper`s version
            val classLiveData = LiveData::class.java
            val fieldObservers = classLiveData.getDeclaredField("mObservers")
            fieldObservers.isAccessible = true
            val objectObservers = fieldObservers.get(this)
            val classObservers = objectObservers.javaClass
            val methodGet = classObservers.getDeclaredMethod("get", Any::class.java)
            methodGet.isAccessible = true
            val objectWrapperEntry = methodGet.invoke(objectObservers, observer)
            var objectWrapper: Any? = null
            if (objectWrapperEntry is Map.Entry<*, *>) {
                objectWrapper = objectWrapperEntry.value
            }
            checkNotNull(objectWrapper) { "Wrapper can not be bull!" }
            val classObserverWrapper = objectWrapper.javaClass.superclass
            val filedLastVersion = classObserverWrapper.getDeclaredField("mLastVersion")
            filedLastVersion.isAccessible = true

            // get livedata`s version
            val fieldVersion = classLiveData.getDeclaredField("mVersion")
            fieldVersion.isAccessible = true
            val objectVersion = fieldVersion.get(this)

            // set wrapper`s version
            filedLastVersion.set(objectWrapper, objectVersion)
        }
    }

    /**
     * 订阅通知
     * @param lifecycleOwner
     */
    fun injectBus(lifecycleOwner: LifecycleOwner) {
        val aClass = lifecycleOwner.javaClass
        var viewLifecycleOwner: LifecycleOwner? = null

        if (lifecycleOwner is Activity) viewLifecycleOwner = lifecycleOwner
        if (lifecycleOwner is Fragment) viewLifecycleOwner = lifecycleOwner.viewLifecycleOwner

        // 获取当前类所有方法
        val declaredMethods = aClass.declaredMethods
        declaredMethods.forEach { method ->
            val injectBus = method.getAnnotation(InjectBus::class.java)
            injectBus?.run {
                check(method.parameterTypes.size == 1) {
                    "$aClass injectBus function params isEmpty or params count > 1."
                }
                val tagType = method.parameterTypes.first()
                val observer = Observer<Any> {
                    try {
                        method.invoke(lifecycleOwner, it)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }

                injectSizeAdd(tagType)

                viewLifecycleOwner?.run {
                    val channel = instance.getChannel(tagType)
                    if (isActive) { // 是否是激活状态 激活状态last粘性消息
                        channel.observe(this, observer)
                    } else { // 非激活状态即时消息
                        channel.observeForever(observer)
                    }

                    // 解除绑定
                    lifecycle.addObserver(object : LifecycleEventObserver {
                        override fun onStateChanged(
                            source: LifecycleOwner,
                            event: Lifecycle.Event
                        ) {
                            if (event == Lifecycle.Event.ON_DESTROY) {
                                // 移除事件订阅
                                channel.removeObserver(observer)
                                // 移除事件
                                injectSizeRemove(tagType)
                                // 移除生命周期监听器
                                source.lifecycle.removeObserver(this)
                            }
                        }
                    })
                }
            }
        }
    }

}